sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java

```diff
@@ -1793,6 +1814,32 @@ static String getExtractDestinationUri(String extractDestinationDir) {
       return toBuilder().setJsonTimePartitioning(partitioning).build();
     }
 
+    /**
+     * Specifies the clustering fields to use when writing to a single output table. Can only be
+     * used when {@link#withTimePartitioning(TimePartitioning)} is set. If {@link
+     * #to(SerializableFunction)} or {@link #to(DynamicDestinations)} is used to write to dynamic
+     * tables, this setting is ignored; instead, see {@link #enableClustering()}.
+     */
+    public Write<T> withClustering(Clustering clustering) {
+      checkArgument(clustering != null, "clustering can not be null");
+      return toBuilder().setEnableClustering(true).setClustering(clustering).build();
+    }
+
+    /**
+     * Allows writing to clustered tables when {@link #to(SerializableFunction)} or {@link
+     * #to(DynamicDestinations)} is used. The returned {@link TableDestination} objects should
+     * specify the clustering fields per table. If writing to a single table, use {@link
+     * #withClustering(Clustering)} instead to specify the clustering fields.
+     *
+     * <p>Setting this option enables use of {@link TableDestinationCoderV3} which encodes
+     * clustering information. Pipelines using an older coder must be drained before setting this
+     * option, since {@link TableDestinationCoderV3} will not be able to read state written with a
+     * previous version.
+     */
+    public Write<T> enableClustering() {
```

Comment:
Having these two methods seems to make the API pretty brittle.

How about just having one function withClustering() that optionally takes a clustering object ? In dynamic destinations case optional Clustering object can be skipped/null and the method will behave similar to enableClustering() ?

Reply:
Implemented. `enableClustering()` is now `withClustering()` and it sets `clustering` to a default `Clustering` instance with no fields set, which we use as a flag for enabling clustering on dynamic destinations.